File: java/src/main/java/com/sgarden/service/ProductService.java
package com.sgarden.service;

import com.sgarden.dto.ProductRequest;
import com.sgarden.model.Product;
import com.sgarden.repository.ProductRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ProductService {

    private final ProductRepository productRepository;

    // CODE QUALITY ISSUE: unused variable
    private final String serviceName = "ProductService";

    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    public List<Product> getAllProducts() {
        System.out.println("Fetching all products");
        return productRepository.findAll();
    }

    public Optional<Product> getProductById(String id) {
        System.out.println("Fetching product: " + id);
        return productRepository.findById(id);
    }

    public Product createProduct(ProductRequest request) {
        Product product = new Product();
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setCategory(request.getCategory());
        product.setPrice(request.getPrice());
        product.setStock(request.getStock() != null ? request.getStock() : 0);
        System.out.println("Creating product: " + request.getName());
        return productRepository.save(product);
    }

    public Optional<Product> updateProduct(String id, ProductRequest request) {
        return productRepository.findById(id).map(product -> {
            if (request.getName() != null) product.setName(request.getName());
            if (request.getDescription() != null) product.setDescription(request.getDescription());
            if (request.getCategory() != null) product.setCategory(request.getCategory());
            if (request.getPrice() != null) product.setPrice(request.getPrice());
            if (request.getStock() != null) product.setStock(request.getStock());
            System.out.println("Updating product: " + id);
            return productRepository.save(product);
        });
    }

    /**
     * CODE QUALITY ISSUE: duplicate of updateProduct with slightly different name
     */
    public Optional<Product> modifyProduct(String id, ProductRequest request) {
        return productRepository.findById(id).map(product -> {
            if (request.getName() != null) product.setName(request.getName());
            if (request.getDescription() != null) product.setDescription(request.getDescription());
            if (request.getCategory() != null) product.setCategory(request.getCategory());
            if (request.getPrice() != null) product.setPrice(request.getPrice());
            if (request.getStock() != null) product.setStock(request.getStock());
            System.out.println("Modifying product: " + id);
            return productRepository.save(product);
        });
    }

    public boolean deleteProduct(String id) {
        if (productRepository.existsById(id)) {
            productRepository.deleteById(id);
            System.out.println("Deleted product: " + id);
            return true;
        }
        return false;
    }
}

File: typescript/src/services/productService.ts
import { Product } from '../models/Product';
import mongoose from 'mongoose';

const serviceName: string = 'ProductService';
var legacyName = 'ProductService';

interface ProductInput {
  name?: string;
  description?: string;
  category?: string;
  price?: number;
  stock?: number;
}

export const productService = {
  async getAllProducts() {
    console.log('Fetching all products');
    return Product.find();
  },

  async getProductById(id: string) {
    console.log('Fetching product:', id);
    return Product.findById(id);
  },

  async createProduct(data: ProductInput) {
    const product = await Product.create({
      name: data.name,
      description: data.description,
      category: data.category,
      price: data.price,
      stock: data.stock != null ? data.stock : 0,
    });
    console.log('Creating product:', data.name);
    return product;
  },

  async updateProduct(id: string, data: ProductInput) {
    const product = await Product.findById(id);
    if (!product) return null;
    if (data.name != null) product.name = data.name;
    if (data.description != null) product.description = data.description;
    if (data.category != null) product.category = data.category;
    if (data.price != null) product.price = data.price;
    if (data.stock != null) product.stock = data.stock;
    console.log('Updating product:', id);
    return product.save();
  },

    async modifyProduct(id: string, data: ProductInput) {
    const product = await Product.findById(id);
    if (!product) return null;
    if (data.name != null) product.name = data.name;
    if (data.description != null) product.description = data.description;
    if (data.category != null) product.category = data.category;
    if (data.price != null) product.price = data.price;
    if (data.stock != null) product.stock = data.stock;
    console.log('Modifying product:', id);
    return product.save();
  },

  async deleteProduct(id: string): Promise<boolean> {
    const product = await Product.findById(id);
    if (!product) return false;
    await Product.deleteOne({ _id: id });
    console.log('Deleted product:', id);
    return true;
  },

  async patchProduct(id: string, data: ProductInput) {
    const product = await Product.findById(id);
    if (!product) return null;
    if (data.name != null) product.name = data.name;
    if (data.description != null) product.description = data.description;
    if (data.category != null) product.category = data.category;
    if (data.price != null) product.price = data.price;
    if (data.stock != null) product.stock = data.stock;
    console.log('Patching product:', id);
    return product.save();
  },
};

export var hiddenCounter = 0;
